smolyay/benchmark/benchmark.py

```diff
@@ -0,0 +1,75 @@
+import abc
+
+import numpy
+
+class BenchmarkFunction(abc.ABC):
+    """Benchmark Function
+
+    Implementation of benchmark problems used for optimization, testing, and 
+    analysis of surrogate function solvers. These benchmark problems come from
+    https://sahinidis.coe.gatech.edu/dfo
+
+    These functions operate on a defined `domain` they can be evaluated on,
+    and the upper and lower bounds of this domain can be the domain in which
+    solutions exist or can be arbitrary.
+
+    """
+    def __call__(self,x):
+        """Evaluate the function.
+
+        Parameters
+        ----------
+        x : list
+            Function input.
+
+        Raises
+        ------
+        ValueError
+            If the input is outside the function domain.
+        """
+        x = numpy.array(x, copy=False,ndmin=2)
+        if self.dimension > 1:
+            oob = any(numpy.any(xi < bound[0]) or numpy.any(xi > bound[1]) 
+                    for xi,bound in zip(x.transpose(),self.domain))
+        else:
+            oob = (numpy.any(x < self.domain[0][0]) or 
+                    numpy.any(x > self.domain[0][1]))
```

Comment:
It actually is because for you're passing multiple inputs for a single dimension, x[...,0] will only pass the first input, not all of the inputs.

Reply:
Oh, I think you are talking about a specific use case where the dimension of the function is 1, and you are allowing a 1d array to be interpreted as multiple inputs? This array structure is different from how the inputs are expected for higher dimensional functions. When cast to a 2d array, it has the wrong dimensionality in the second index:
```
>>> x = np.array([1, 2, 3], ndmin=2)
>>> x
array([[1, 2, 3]])
>>> x.shape
(1, 3)
```
Whereas we would expect it to have the transposed shape if we explicitly interpreted this as an array of 1d inputs:
```
>>> y = np.array([[1],[2],[3]], ndmin=2)
>>> y
array([[1],
       [2],
       [3]])
>>> y.shape
(3, 1)
```
You could check for this special case (`self.dimension == 1 and x.shape[0]  > 1`), and transpose x before the function proceeds.

Otherwise, it is very unclear why you would do it this way.

Reply:
If the case where dimension is 1 is still treated as a special case, then the if-else is just being replaced with a different if statement, which kind of defeats the whole purpose of making the if-else statement unnecessary.

Reply:
This is not just the 1d case, it is more specific than that. It reduces code duplication and makes the purpose of the code clearer & more consistent, since you will sanitize `x` to have the right shape in this corner case that you are allowing, which is different from all of the other use cases. Hence, subsequent usage of `x` in the function will be more unified.

So no, it does not

> defeats the whole purpose of making the if-else statement unnecessary

Reply:
Okay, it is now done.

Reply:
As an aside, in the surrogate pull request that allows multiple inputs for call and gradient, I used a similar bounds checking method. However, surrogate allows for the domain to be a single list, so using self.domain[i][0] wouldn't work for the 1D case in the Surrogate class.